xbmc/platform/linux/input/LibInputHandler.cpp

```diff
@@ -0,0 +1,249 @@
+    CLog::Log(LOGERROR, "%s - failed to open %s (%s)", __FUNCTION__, path, strerror(errno));
+    return -errno;
+  }
+
+  auto ret = ioctl(fd, EVIOCGRAB, (void*)1);
+  if (ret < 0)
+  {
+    CLog::Log(LOGDEBUG, "%s - grab requested, but failed for %s (%s)", __FUNCTION__, path, strerror(errno));
+  }
+
+  return fd;
+}
+
+static void close_restricted(int fd, void *user_data)
+{
+  (void) user_data;
+  close(fd);
+}
+
+static const struct libinput_interface m_interface =
+{
+  open_restricted,
+  close_restricted,
+};
+
+static void LogHandler(libinput *libinput, libinput_log_priority priority, const char *format, va_list args)
+{
+  (void)libinput;
+
+  if (priority == LIBINPUT_LOG_PRIORITY_DEBUG)
+  {
+    char buf[512];
+    int n = vsnprintf(buf, sizeof(buf), format, args);
+    if (n > 0)
+      CLog::Log(LOGDEBUG, "libinput: %s", buf);
+  }
+}
+
+CLibInputHandler::CLibInputHandler(CWinEventsLinux *winEvents)
+{
+  m_udev = udev_new();
+  if (!m_udev)
+  {
+    CLog::Log(LOGERROR, "CLibInputHandler::%s - failed to get udev context for libinput", __FUNCTION__);
+    return;
+  }
+
+  m_li = libinput_udev_create_context(&m_interface, nullptr, m_udev);
+  if (!m_li)
+  {
+    CLog::Log(LOGERROR, "CLibInputHandler::%s - failed to get libinput context", __FUNCTION__);
+    return;
+  }
+
+  libinput_log_set_handler(m_li, LogHandler);
+  libinput_log_set_priority(m_li, LIBINPUT_LOG_PRIORITY_DEBUG);
+
+  auto ret = libinput_udev_assign_seat(m_li, "seat0");
+  if (ret < 0)
+    CLog::Log(LOGERROR, "CLibInputHandler::%s - failed to assign seat", __FUNCTION__);
+
+  m_liFd = libinput_get_fd(m_li);
+
+  m_keyboard.reset(new CLibInputKeyboard(winEvents));
+  m_pointer.reset(new CLibInputPointer(winEvents));
+  m_touch.reset(new CLibInputTouch());
+}
+
+CLibInputHandler::~CLibInputHandler()
+{
+  if (m_li)
+    libinput_unref(m_li);
+
+  if (m_udev)
+    udev_unref(m_udev);
+}
+
+void CLibInputHandler::OnReadyRead()
+{
+  auto ret = libinput_dispatch(m_li);
+  if (ret < 0)
+  {
+    CLog::Log(LOGERROR, "CLibInputHandler::%s - libinput_dispatch failed", __FUNCTION__);
+    return;
+  }
+
+  libinput_event *ev;
+  while ((ev = libinput_get_event(m_li)) != nullptr)
+  {
+    ProcessEvent(ev);
+    libinput_event_destroy(ev);
+  }
+}
+
+void CLibInputHandler::ProcessEvent(libinput_event *ev)
+{
+  libinput_event_type type = libinput_event_get_type(ev);
+  libinput_device *dev = libinput_event_get_device(ev);
+
+  switch (type)
+  {
+    case LIBINPUT_EVENT_DEVICE_ADDED:
+    {
+      const char *sysname = libinput_device_get_sysname(dev);
+      const char *name = libinput_device_get_name(dev);
+
+      if (libinput_device_has_capability(dev, LIBINPUT_DEVICE_CAP_TOUCH))
+      {
+        auto device = std::find(m_devices.begin(), m_devices.end(), dev);
+        if (device == m_devices.end())
+        {
+          CLog::Log(LOGDEBUG, "CLibInputHandler::%s - touch type device added: %s (%s)", __FUNCTION__, name, sysname);
+          m_devices.push_back(dev);
```

Comment:
you need a `libinput_device_ref()` here, otherwise the lifetime isn't guaranteed.